test/distributed/_shard/sharding_plan/test_sharding_plan.py

```diff
@@ -0,0 +1,149 @@
+from torch.distributed._shard.sharding_plan import ShardingPlan, ShardingPlanner
+from torch.distributed._shard.sharding_spec import ChunkShardingSpec
+from torch.distributed._shard.sharded_tensor import ShardedTensor
+
+from torch.testing._internal.common_utils import TEST_WITH_DEV_DBG_ASAN
+from torch.testing._internal.distributed._shard.sharded_tensor import (
+    TEST_GPU_NUM,
+    ShardedTensorTestBase,
+    with_comms,
+)
+from torch.testing._internal.distributed._shard.sharded_tensor._test_ops_common import (
+    generate_chunk_sharding_specs_for_test,
+)
+
+if TEST_WITH_DEV_DBG_ASAN:
+    print(
+        "Skip dev-asan as torch + multiprocessing spawn have known issues",
+        file=sys.stderr,
+    )
+    sys.exit(0)
+
+
+class SimpleMegatronLM(nn.Module):
+    def __init__(self, linear_size, rank=None):
+        super().__init__()
+        self.fc1 = nn.Linear(*linear_size[0])
+        self.gelu = nn.GELU()
+        self.fc2 = nn.Linear(*linear_size[1])
+        if rank:
+            self.fc1.cuda(rank)
+            self.fc2.cuda(rank)
+
+    def forward(self, inp):
+        return self.fc2(self.gelu(self.fc1(inp)))
+
+# Example ShardingPlanner that chunks every parameter in the module
+# to all available devices defined.
+class ChunkAllShardingPlanner(ShardingPlanner):
+    dim = 0
+    devices = []
+
+    def __init__(self, chunk_dim=0, device_count=0):
+        self.dim = chunk_dim
+        self.devices = [f"rank:{i}/cuda:{i}" for i in range(device_count)]
+
+    def build_plan(self, module: nn.Module) -> ShardingPlan:
+        named_params = module.named_parameters()
+        plan = {}
+        for name, param in named_params:
+            plan[name] = ChunkShardingSpec(self.dim, placements=self.devices)
+
+        return ShardingPlan(plan=plan)
+
+class TestShardingPlan(ShardedTensorTestBase):
+
+    @with_comms(init_rpc=False)
+    @skip_if_lt_x_gpu(TEST_GPU_NUM)
+    @requires_nccl()
+    def test_basic_sharding_plan(self):
+        # since we already test the numerics in test_megatron_prototype.py
+        # this test is only testing the ShardingPlan and whether the model
+        # be sharded with the ShardingPlan, we don't need to verify the
+        # numeric correctness.
+        colwise_sharding_spec = generate_chunk_sharding_specs_for_test(0)
+        rowwise_sharding_spec = generate_chunk_sharding_specs_for_test(1)
+        for spec in zip(colwise_sharding_spec, rowwise_sharding_spec):
+            # test each sharding spec pair and see if we can apply sharding
+            sharding_plan = ShardingPlan(plan={
+                "fc1.weight": spec[0],
+                "fc2.weight": spec[1],
+            })
+
+            megatron_lm = SimpleMegatronLM([[17, 12], [12, 29]], rank=self.rank).cuda(
+                self.rank
+            )
+
+            # shard the module with the provided sharding plan
+            shard_module(megatron_lm, sharding_plan)
```

Comment:
Don't we need to `reshard` back to dataparallel after fc2? We need a way to specify that as well.

Reply:
Yep just updated the PR to support `reshard_output` + `collect_local_shards`, all documented in the `ShardingPlan` class, let me know if this make sense.